### PR TITLE
bpo-46642: explicitly disallow typing.TypeVar instance subclassing

### LINKED ISSUE #90800: typing: tested TypeVar instance subclass TypeError is incidental
BPO | [46642](https://bugs.python.org/issue46642)
--- | :---
Nosy | @serhiy-storchaka, @JelleZijlstra, @sobolevn, @Fidget-Spinner, @AlexWaygood, @GBeauregard
PRs | <li>python/cpython#31148</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-02-04.21:59:21.151>
labels = ['type-feature', 'library']
title = 'typing: tested TypeVar instance subclass TypeError is incidental'
updated_at = <Date 2022-03-12.03:39:46.337>
user = 'https://github.com/GBeauregard'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-12.03:39:46.337>
actor = 'JelleZijlstra'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2022-02-04.21:59:21.151>
creator = 'GBeauregard'
dependencies = []
files = []
hgrepos = []
issue_num = 46642
keywords = ['patch']
message_count = 7.0
messages = ['412544', '412589', '412591', '412627', '412646', '412661', '414970']
nosy_count = 6.0
nosy_names = ['serhiy.storchaka', 'JelleZijlstra', 'sobolevn', 'kj', 'AlexWaygood', 'GBeauregard']
pr_nums = ['31148']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue46642'
versions = []
```

</p></details>

### PR DESCRIPTION
**Pull request by GBeauregard**
The existing test covering this case passed only incidentally. We
explicitly disallow doing this and add a proper error message.

The way [this test](https://github.com/python/cpython/blob/bf95ff91f2c1fc5a57190491f9ccdc63458b089e/Lib/test/test_typing.py#L227-L230) passed before this PR is almost certainly a bug (see the bpo issue)

https://bugs.python.org/issue46642

Closes  #90800.

### CODE REVIEW

**Comment by GBeauregard** [comment]
The failed test is because of [bpo-46648](https://bugs.python.org/issue46648) and not related to this PR.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
To make the test more precise, maybe add an `assertRaisesRegex` with the error message you add here.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by gvanrossum** [comment]
On bpo, @serhiy-storchaka seems to argue that the original code is correct so this should not be merged. I will make him a reviewer so he can merge if he agrees with the change; if not, I propose that we let the status quo prevail and close both PR and issue without action.

**Comment by GBeauregard** [comment]
Okay. @serhiy-storchaka we can probably resolve #30268 at the same time.

If we decide against these error messages we should be able to clean up other instances them in typing.py as well.

**Comment by Fidget-Spinner** [comment]
> On bpo, @serhiy-storchaka seems to argue that the original code is correct so this should not be merged. I will make him a reviewer so he can merge if he agrees with the change; if not, I propose that we let the status quo prevail and close both PR and issue without action.

It seems to me that he proposes keeping the test but that the new `__mro_entries__` code is redundant. If so, I concur and I'll be happy to merge this with just the test.

File: Lib/typing.py [hunk 1]
```diff
@@ -818,6 +818,9 @@ def __init__(self, name, *constraints, bound=None,
         if def_mod != 'typing':
             self.__module__ = def_mod
 
+    def __mro_entries__(self, bases):
```
Comment (serhiy-storchaka) [inline Lib/typing.py:1107]: It should be added not here, but in `_TypeVarLike`. `ParamSpec` instances are not subclassable too.

**Comment by serhiy-storchaka** [comment]
For now, adding `__mro_entry__` is not necessary. Although it would not harm too.

File: Lib/typing.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/typing.py:1107]: And add it also in TypeVarTuple.

Comment (GBeauregard) [reply Lib/typing.py:1107]: Thanks for the feedback! I'll get to this soon and ping you when done (I'm currently in the process of moving across the country and starting a job)

**Pull request merged**